include/deal.II/fe/fe_coupling_values.h

```diff
@@ -406,6 +407,610 @@ namespace FEValuesViews
+ *
+ * Internally, this class provides an abstraction to organize coupling terms
+ * between two arbitrary FEValuesBase objects, and provides a unified way to
+ * access the values of the two bases, and of the two sets of quadrature points
+ * and weights.
+ *
+ * Depending on the DoFCouplingType and QuadratureCouplingType argument passed
+ * to the constructor, the class behaves differently w.r.t. how the dof
+ * indices and the quadrature points of the two different FEValuesBase objects
+ * are grouped together.
+ *
+ * The class is intended to be a higher level abstraction compared to assembling
+ * coupling terms manually.
+ *
+ * This class gives access to two different FEValuesBase objects, and it
+ * provides a unified way to access the values of the two basis, using the
+ * concept of FEValuesExtractors. Unlike the FEValuesBase class, though, we need
+ * to specify which FEValuesBase object we are referring to, and this is done by
+ * calling the first() or second() functions, which are used to *inform* the
+ * extractors about what FEValuesBase object the extractor refers to.
+ *
+ * This is achieved, in user codes, by the following snippet:
+ *
+ * @code
+ * // extract the first scalar component of the basis
+ * FEValuesExtractor scalar(0);
+ * ...
+ *
+ * FECouplingValues<dim> fe_coupling(fev1, fev1);
+ *
+ * // Extractors for the two FEValuesBase objects are returned by the first()
+ * // and second() methods
+ *
+ * const auto first = fe_coupling.first(scalar);
+ * const auto second = fe_coupling.second(scalar);
+ *
+ * // Now we can use the augmented extractors to access the values of the two
+ * // FEValuesBase objects
+ * const auto & first_vi = fe_coupling[first].value(i, q);
+ * const auto & second_vi = fe_coupling[second].value(i, q);
+ * @endcode
+ */
+template <int dim1, int dim2 = dim1, int spacedim = dim1>
+class FECouplingValues
+{
+public:
+  /**
+   * Construct the FECouplingValues in an invalid state. Before you can use this
+   * object, you must call the reinit() function.
+   */
+  FECouplingValues();
+
+  /**
+   * Construct the FECouplingValues with two arbitrary FEValuesBase objects.
+   * This class assumes that the FEValuesBase objects that are given at
+   * construction time are initialized and ready to use (i.e., that you have
+   * called the reinit() function on them before calling this constructor).
+   *
+   * @param fe_values_1 The first FEValuesBase object.
+   * @param fe_values_2 The second FEValuesBase object.
+   * @param dof_coupling_type The type of dof coupling to use.
+   * @param quadrature_coupling_type The type of quadrature to use for the coupling.
+   */
+  FECouplingValues(
+    const FEValuesBase<dim1, spacedim> &fe_values_1,
+    const FEValuesBase<dim2, spacedim> &fe_values_2,
+    const DoFCouplingType &dof_coupling_type = DoFCouplingType::independent,
+    const QuadratureCouplingType &quadrature_coupling_type =
+      QuadratureCouplingType::tensor_product);
```

Comment:
This really depends on the application and on the specific type of coupling. For example, for volume/volume coupling, i.e., for operators with non-local (and non-singular) kernels of type 
$$
\int_K \int_T  f(\phi_i(x)-\phi_j(y), x-y) dx dy
$$
you may initialize `FECouplingValues` once, and just reinit the underlying `FEValuesBase` objects on different cells `K` and `T`, without the need to recompute the coupling (i.e., the numbering is always the same, and nothing differs from what happened in the first call).

For cell/surface coupling, the same cell may couple with different faces, so the renumbering must be really computed from scratch for each pair of `FEValuesBase` objects, so reinitializing the underlying cells and faces will make the renumbering itself invalid, and  `FECouplingValues` must be reinitialized (o constructed from scratch) after calling `fev1->reinit(K)` and `fev2->reinit(T)`.

Reply:
I see. Do you think you could put something like this in the documentation? I think this would be valuable for users to see -- basically just copy-pasting what you have would do.